File: rust/cmd/doubaospeech/src/commands/translation.rs
//! Translation (simultaneous interpretation) CLI commands

use clap::{Args, Subcommand};
use serde::{Deserialize, Serialize};

use giztoy_doubaospeech::{
    AudioFormat, Language, SampleRate, TranslationAudioConfig, TranslationConfig,
};

use super::{create_client, get_context, load_request, output_result, print_success, print_verbose};
use crate::Cli;

/// Simultaneous translation service
#[derive(Args)]
pub struct TranslationCommand {
    #[command(subcommand)]
    command: TranslationSubcommand,
}

#[derive(Subcommand)]
enum TranslationSubcommand {
    /// Stream simultaneous translation
    Stream {
        /// Audio file path
        #[arg(short = 'a', long)]
        audio: Option<String>,
        /// Source language
        #[arg(short = 's', long, default_value = "zh-CN")]
        source_lang: String,
        /// Target language
        #[arg(short = 't', long, default_value = "en-US")]
        target_lang: String,
        /// Enable TTS output
        #[arg(long)]
        enable_tts: bool,
    },
    /// Interactive translation mode
    Interactive {
        /// Source language
        #[arg(short = 's', long, default_value = "zh-CN")]
        source_lang: String,
        /// Target language
        #[arg(short = 't', long, default_value = "en-US")]
        target_lang: String,
    },
}

/// Translation config from YAML/JSON file
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[allow(dead_code)]
struct TranslationFileConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    audio_file: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    source_language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    target_language: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    sample_rate: Option<i32>,
}

impl TranslationCommand {
    pub async fn run(&self, cli: &Cli) -> anyhow::Result<()> {
        match &self.command {
            TranslationSubcommand::Stream {
                audio,
                source_lang,
                target_lang,
                enable_tts,
            } => {
                self.stream(cli, audio.as_deref(), source_lang, target_lang, *enable_tts)
                    .await
            }
            TranslationSubcommand::Interactive {
                source_lang,
                target_lang,
            } => self.interactive(cli, source_lang, target_lang).await,
        }
    }

    async fn stream(
        &self,
        cli: &Cli,
        audio_path: Option<&str>,
        source_lang: &str,
        target_lang: &str,
        enable_tts: bool,
    ) -> anyhow::Result<()> {
        let ctx = get_context(cli)?;
        let client = create_client(&ctx)?;
        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(
            cli,
            &format!("Source: {} -> Target: {}", source_lang, target_lang),
        );

        // Load audio file
        let audio_file = if let Some(input_file) = cli.input.as_deref() {
            let file_cfg: TranslationFileConfig = load_request(input_file)?;
            audio_path
                .map(|s| s.to_string())
                .or(file_cfg.audio_file)
                .ok_or_else(|| anyhow::anyhow!("audio file is required"))?
        } else {
            audio_path
                .ok_or_else(|| anyhow::anyhow!("audio file is required, use -a flag"))?
                .to_string()
        };

        let audio_data = std::fs::read(&audio_file)?;
        print_verbose(
            cli,
            &format!("Audio file: {} ({} bytes)", audio_file, audio_data.len()),
        );

        // Parse languages
        let source_language = match source_lang {
            "zh-CN" => Language::ZhCn,
            "en-US" => Language::EnUs,
            "ja-JP" => Language::JaJp,
            _ => Language::ZhCn,
        };
        let target_language = match target_lang {
            "zh-CN" => Language::ZhCn,
            "en-US" => Language::EnUs,
            "ja-JP" => Language::JaJp,
            _ => Language::EnUs,
        };

        let config = TranslationConfig {
            source_language,
            target_language,
            audio_config: TranslationAudioConfig {
                format: AudioFormat::Pcm,
                sample_rate: SampleRate::Rate16000,
                channel: 1,
                bits: 16,
            },
            enable_tts,
            tts_voice: None,
        };

        print_verbose(cli, "Opening translation session...");
        let session = client.translation().open_session(&config).await?;

        // Send audio in chunks
        let chunk_size = 3200; // 100ms of 16kHz 16-bit mono audio
        let chunks: Vec<&[u8]> = audio_data.chunks(chunk_size).collect();
        let total_chunks = chunks.len();

        for (i, chunk) in chunks.iter().enumerate() {
            let is_last = i == total_chunks - 1;
            session.send_audio(chunk, is_last).await?;
            if !is_last {
                tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
            }
        }
        print_verbose(cli, &format!("Sent {} audio chunks", total_chunks));

        // Receive results
        let mut translations = Vec::new();
        while let Some(result) = session.recv().await {
            match result {
                Ok(chunk) => {
                    if !chunk.source_text.is_empty() || !chunk.target_text.is_empty() {
                        print_verbose(
                            cli,
                            &format!("{} -> {}", chunk.source_text, chunk.target_text),
                        );
                        translations.push(serde_json::json!({
                            "source": chunk.source_text,
                            "target": chunk.target_text,
                            "is_final": chunk.is_final,
                        }));
                    }
                    if chunk.is_final {
                        break;
                    }
                }
                Err(e) => {
                    return Err(anyhow::anyhow!("Translation error: {}", e));
                }
            }
        }

        session.close().await?;
        print_success("Translation completed");

        let result = serde_json::json!({
            "translations": translations,
        });
        output_result(&result, cli.output.as_deref(), cli.json)
    }

    async fn interactive(
        &self,
        cli: &Cli,
        source_lang: &str,
        target_lang: &str,
    ) -> anyhow::Result<()> {
        let ctx = get_context(cli)?;
        print_verbose(cli, &format!("Using context: {}", ctx.name));

        // Interactive mode requires microphone input which is complex
        // For now, provide guidance
        eprintln!("Interactive translation mode requires microphone input.");
        eprintln!("Use 'stream' subcommand with an audio file instead:");
        eprintln!(
            "  doubaospeech translation stream -a audio.pcm -s {} -t {}",
            source_lang, target_lang
        );

        let result = serde_json::json!({
            "note": "Use 'stream' subcommand with audio file",
            "source_language": source_lang,
            "target_language": target_lang,
        });
        output_result(&result, cli.output.as_deref(), cli.json)
    }
}

File: go/cmd/doubaospeech/commands/translation.go
package commands

import (
	"bytes"
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	ds "github.com/haivivi/giztoy/go/pkg/doubaospeech"
)

var translationCmd = &cobra.Command{
	Use:   "translation",
	Short: "Simultaneous translation service",
	Long: `Simultaneous translation service.

Real-time speech-to-speech translation.

Example config file (translation.yaml):
  source_language: zh-CN
  target_language: en-US
  audio_config:
    format: pcm
    sample_rate: 16000
    channel: 1
    bits: 16
  enable_tts: true
  tts_voice: en_female_sweet`,
}

var translationStreamCmd = &cobra.Command{
	Use:   "stream",
	Short: "Stream simultaneous translation",
	Long: `Stream simultaneous translation.

Send audio in the source language and receive
translated audio in the target language.

Example config file (translation.yaml):
  source_language: zh-CN
  target_language: en-US
  enable_tts: true
  tts_voice: en_female_sweet

Examples:
  doubaospeech -c myctx translation stream -f translation.yaml --audio input.pcm -o output.pcm`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		audioFile, err := cmd.Flags().GetString("audio")
		if err != nil {
			return fmt.Errorf("failed to read 'audio' flag: %w", err)
		}
		if audioFile == "" {
			return fmt.Errorf("--audio flag is required for streaming translation")
		}

		cliCtx, err := getContext()
		if err != nil {
			return err
		}

		client, err := createClient(cliCtx)
		if err != nil {
			return err
		}

		var config ds.TranslationConfig
		if err := loadRequest(getInputFile(), &config); err != nil {
			return err
		}

		printVerbose("Using context: %s", cliCtx.Name)
		printVerbose("Source language: %s", config.SourceLanguage)
		printVerbose("Target language: %s", config.TargetLanguage)

		reqCtx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
		defer cancel()

		return runTranslationStream(reqCtx, client, &config, audioFile)
	},
}

var translationInteractiveCmd = &cobra.Command{
	Use:   "interactive",
	Short: "Interactive translation mode",
	Long: `Start an interactive translation session.

This mode captures audio from your microphone and plays
translated audio through your speakers.

Examples:
  doubaospeech -c myctx translation interactive -f translation.yaml`,
	RunE: func(cmd *cobra.Command, args []string) error {
		return fmt.Errorf("not implemented: interactive translation mode requires microphone/speaker hardware integration; use 'doubaospeech translation stream' for file-based translation")
	},
}

// ============================================================================
// Implementation Functions
// ============================================================================

func runTranslationStream(ctx context.Context, client *ds.Client, config *ds.TranslationConfig, audioFile string) error {
	session, err := client.Translation.OpenSession(ctx, config)
	if err != nil {
		return fmt.Errorf("failed to open translation session: %w", err)
	}
	defer session.Close()

	printVerbose("Translation session opened")

	// Read audio data
	audioData, err := readAudioInput(audioFile)
	if err != nil {
		return fmt.Errorf("failed to read audio: %w", err)
	}

	printVerbose("Sending audio (%s)...", formatBytes(int64(len(audioData))))

	// Send audio in chunks
	chunkSize := 3200 // 100ms of 16kHz 16-bit mono
	for i := 0; i < len(audioData); i += chunkSize {
		end := i + chunkSize
		isLast := end >= len(audioData)
		if isLast {
			end = len(audioData)
		}

		if err := session.SendAudio(ctx, audioData[i:end], isLast); err != nil {
			return fmt.Errorf("send audio: %w", err)
		}

		if !isLast {
			time.Sleep(100 * time.Millisecond)
		}
	}

	// Receive results
	outputPath := getOutputFile()
	var audioBuf bytes.Buffer
	var finalSource, finalTarget string

	for chunk, err := range session.Recv() {
		if err != nil {
			return fmt.Errorf("receive error: %w", err)
		}

		if chunk.SourceText != "" {
			if chunk.IsDefinite {
				printInfo("[Source] %s", chunk.SourceText)
			} else {
				printVerbose("[Source interim] %s", chunk.SourceText)
			}
			finalSource = chunk.SourceText
		}
		if chunk.TargetText != "" {
			if chunk.IsDefinite {
				printInfo("[Target] %s", chunk.TargetText)
			} else {
				printVerbose("[Target interim] %s", chunk.TargetText)
			}
			finalTarget = chunk.TargetText
		}
		if len(chunk.Audio) > 0 {
			audioBuf.Write(chunk.Audio)
			printVerbose("[Audio] +%d bytes", len(chunk.Audio))
		}
		if chunk.IsFinal {
			break
		}
	}

	if audioBuf.Len() > 0 && outputPath != "" {
		if err := outputBytes(audioBuf.Bytes(), outputPath); err != nil {
			return fmt.Errorf("write audio: %w", err)
		}
		printSuccess("Translated audio saved to: %s (%s)", outputPath, formatBytes(int64(audioBuf.Len())))
	}

	result := map[string]any{
		"source_text": finalSource,
		"target_text": finalTarget,
	}
	if audioBuf.Len() > 0 {
		result["audio_size"] = audioBuf.Len()
	}

	return outputResult(result, "", isJSONOutput())
}

func init() {
	translationStreamCmd.Flags().String("audio", "", "Audio file path (required)")

	translationCmd.AddCommand(translationStreamCmd)
	translationCmd.AddCommand(translationInteractiveCmd)
}
